Accommodate duplicate and empty world names automatically

metarmask opened the pull request:
When the create button is pressed and the name is empty it becomes "Unnamed" and if duplicate a number is added or increased.

Commit list:

- ~~Start world after creation~~
- Accommodate world name automatically
- Refactor and fix no escaping in dlg_create_world 
  - Fix game names and map generators not being escaped
  - Fix variable name inconsistencies within file
  - Use table concatenation to build comma separated lists
  - Use tables for form values

sofar commented:
If you change a visual aspect, especially menus, it helps to show screenshots of before-and-after in the PR thread.

Ferk commented:
> * Start world after creation
> * Change world name automatically if duplicate or empty

Cool, I'm tired of having to create random "asdfad" world names.

However, the combination of these two makes it hard to know what's the final name of the game that was just created, because when it automatically starts you don't see the name that was used.

Maybe the name of the game should be displayed somewhere while it's loading.. or maybe a dialog should warn when the name you wrote was automatically altered/chosen.

Also, not sure if appending "_" every time is the best way. You cannot easily tell apart "Unnamed____" and "Unnamed_____", best would be to parse the last character as number, if valid increment it, else append "_1".

Also, it would have been cleaner to have the refactor in a different commit.

Ferk commented:
> If you change a visual aspect, especially menus, it helps to show screenshots of before-and-after in the PR thread.

I don't think there's much of a visual change here.

metarmask commented:
> It would have been cleaner to have the refactor in a different commit, too.

> But yes, the PR can be confusing, it would have been better if the refactor was separate.

I've split the commit.

> Also, not sure if appending "\_" every time is the best way. You cannot easily tell apart "Unnamed\_\_\_\_" and "Unnamed\_\_\_\_\_", best would be to parse the last space-separated word as number, if valid increment it, else append " 1".

You're right, I have changed that.

> However, the combination of these two makes it hard to know what's the final name of the game that was just created, because when it automatically starts you don't see the name that was used.
>
> Maybe the name of the game should be displayed somewhere while it's loading.. or maybe a dialog should warn when the name you wrote was automatically altered/chosen.

I thought of this too. I wasn't able to find a way to update a label which is what I believe would be the best option. But displaying nothing at all may also work as finding the world name with the highest number is easy. "Unnamed" may be a bit harder but not too hard, I believe.

e-y-e commented:
Please correct me if I am wrong and I apologise for discussing the idea and not the implementation, but surely starting the world after creation blocks the opportunity to add mods that change the world gen?

metarmask commented:
> starting the world after creation blocks the opportunity to add mods that change the world gen

@e-y-e I had no clue mods had to be enabled that way until now. I'll try to add the "configure" button to the dialog.

sfan5 commented on builtin/mainmenu/dlg_create_world.lua:
```diff
@@ -15,115 +15,164 @@
+			dialog:show()
+		elseif
+		           this.parent
+		       and this.parent.current_tab == "singleplayer"
```
should be indented with tabs, also the `this.parent` can fit after the `elseif`

nerzhul commented:
can you rebase it to let us a chance to merge it before feature freeze ?

paramat commented:
Please update the first post with details of how 'configure' now works, and screenshots instead of a far-too-fast GIF.

metarmask commented:
@nerzhul The commits have been rebased.

@paramat The post has been updated. I didn't think the speed of the GIF would be that bad as the viewer would know where to focus after reading the title.

paramat commented:
:-1: I like the avoiding of duplicate names and the auto-creation of name if blank, but disagree with starting a world automatically after creation. I know that players sometimes like to create a world, then exit MT to edit the map_meta.txt before entering that world. I find it useful to be able to create a world without automatically entering it. There are very likely other good reasons for how it acts currently.

metarmask commented:
@paramat Sorry for being unclear when explaining the new button. The original plan I had was to start the world automatically after the configure dialog was closed. It failed due to me not finding an easy way to do so. That failure means the button allows what you mentioned.

But all these hidden options... I did not have the slightest clue they even existed before I made this pull request. What "good reason" is there to hide them? Rather than using them as justifications for not implementing this feature, I would use _this_ change as a justification - make them all accessible in one place where they can be discovered by anyone.

hkzorman commented:
I will not comment on the other features, but please __do not__ start world automatically.

Usually, I create a world, then activate mods, then start it. Some mods, especially mapgen mods, doesn't really work well (this is by experience) if not activated before first run. For example, mg_villages mod. I have difficulty finding villages if I don't activate the mod before the very first run of the world.
Plus, what is mentioned by @paramat  above. I also do this, set things in minetest.conf (like mapgen paramaters, etc.) before running world.

nerzhul commented:
does adding a configuration variable to start world automatictly fit your needs ? @metarmask can you add it ?

paramat commented:
> That failure means the button allows what you mentioned.

So the world is not automatically entered after creation? To be clear i'm saying the world should not be entered automatically after creation whether you configure it or not.

> But all these hidden options... I did not have the slightest clue they even existed before I made this pull request. What "good reason" is there to hide them? Rather than using them as justifications for not implementing this feature, I would use this change as a justification - make them all accessible in one place where they can be discovered by anyone.

They are not hidden options, and they cannot be made accessible through code changes, i'm describing various ways of using MT made possible by the current behaviour which would then not become possible. The feature of entering a world automatically after creation saves maybe one mouse click, but you are suggesting we instead attempt to write a lot of complex code to change MT just for the sake of saving one mouse click, there is no justification and we are very understaffed and under-resourced (people don't seem to realise how little dev time we have, and that's why we have to say no so often).

paramat commented:
Please no configuration for auto start, it's extra code we have to write and maintain just for saving one mouse click, the work is not justified.

hkzorman commented:
@nerzhul Yes, a configuration variable would help. 
We could also, instead of configuration, add a button that says "Run" or "Play now" for those who want to start automatically.

paramat commented:
No one has complained about the current behaviour, even the first post assumes that other players want this. Let's not waste precious dev time on unnecessary things.

metarmask commented:
> So the world is not automatically entered [...]? [...] i'm saying the world should not be entered [...] whether you configure it or not.

Not if you click the "Create & configure" button (nor in the server tab). What's wrong with starting the world automatically in some cases if it doesn't break anything?

> but you are suggesting we instead attempt to write a lot of complex code [...] just for the sake of saving one mouse click

Now I'm confused... I've already written the code? I was trying to say that, instead of not adding this feature because there are some options that are set in an non-intuitive way, we make those options intuitive to set... eventually.

It would be a shame if I wasted time on a feature nobody wants. The impression I've gotten in the month since this pull request was created is that at least three people want it.

> They are not hidden options [...]

What I meant is that they're not anywhere in the GUI and therefore can't be discovered by a casual user.

paramat commented:
> What's wrong with starting the world automatically in some cases if it doesn't break anything?

Mods almost always have to be enabled when creating a new world, so automatic entry will not be used much.

> I've already written the code?

I was referring to the work needed to implement this:

> make them all accessible in one place where they can be discovered by anyone.
... we make those options intuitive to set... eventually.

However as i explained, you misunderstand, the various things made possible by current behaviour are not options that can be accessed in one place. if someone needs to access map_meta.txt to edit it before entering a world that can only be done by using a text editor external to MT.

> It would be a shame if I wasted time on a feature nobody wants. The impression I've gotten in the month since this pull request was created is that at least three people want it.

Don't worry it's not at all wasted. The code cleanup and world name features are much appreciated. The automatic entry to worlds after 'create' is just a small part of this.

metarmask commented:
> Mods almost always have to be enabled when creating a new world

Oh... I thought subgames and mods included with a world was the most common way because that's basically the only things I've used. When I wanted to install a standalone mod I was confused when they "didn't work" because I didn't know you had to activate them.

> various things [possible now cannot] be accessed in one place. [Accessing map_meta.txt] can only be done by using a text editor external to MT.

Couldn't you generate that file using a GUI? Is it not worth making some features accessible? I know you think it is a waste of time but isn't it also a waste of time implementing options that will only be used by few players because they're hard to discover?

> Don't worry it's not at all wasted. The code cleanup and world name features are much appreciated

Oh yeah... Those were also in this pull request. Thanks. I'll remove the "start world automatically" commit.

paramat commented:
> Couldn't you generate that file using a GUI? Is it not worth making some features accessible? [...] isn't it also a waste of time implementing options that will only be used by few players because they're hard to discover?

That would be a lot of work, and editing map_meta.txt is not officially supported or recommended, and should not be, however some still like to alter it manually. Also, that's not the only issue.

To be clear i support the code cleanups and the autocreation of world names, i will find that useful when i am rapidly creating test worlds. Removing my -1.

ShadowNinja commented:
Related: #4962

paramat commented:
I support concept.

rubenwardy commented:
:-1: it's not that hard to mix this in with another solution

rubenwardy commented on builtin/mainmenu/dlg_create_world.lua:
```diff
@@ -15,115 +15,148 @@
+		values.name = fgettext_ne("Unnamed")
+	end
+	while menudata.worldlist:uid_exists_raw(values.name) do
+		local words = {}
```
this looks like a hack to me

paramat commented:
I suspect the delay on this is because the very welcome 2nd commit is based on a refactor, we are short on dev time and refactors are hard work to review. Maybe split the 2nd commit off into a separate PR?

paramat commented on builtin/mainmenu/dlg_create_world.lua:
```diff
@@ -102,20 +102,31 @@ local function create_world_button_handler(this, fields)
+		end
+		local number = tonumber(words[#words])
+		if number == nil then
+			values.name = values.name .. " 2"
```
Not keen on a space, as the space will appear in the world filename and spaces in filenames are not good.

metarmask replied:
But then it looks ugly. You should have been there whenever the code that sets the filename as the real name was reviewed. Oh well. I don't have a dev environment setup at the moment, can't you edit the pull request for me?

paramat replied:
No need, the 2nd commit is being resubmitted in #6257

paramat commented on builtin/mainmenu/dlg_create_world.lua:
```diff
@@ -102,20 +102,31 @@ local function create_world_button_handler(this, fields)
+			values.name = values.name .. " 2"
+		else
+			words[#words] = tostring(number + 1)
+			values.name = table.concat(words, " ")
```
I prefer no space.

paramat commented:
Not tested but the 2nd commit seems good, apart from my line comments, and i approve the concept.

paramat commented:
2nd commit re-submitted in #6257

paramat commented:
#6257 merged.
